<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Converts a (list of) bioconductor assay containers into a FacileDataSet. — as.FacileDataSet • FacileData</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Converts a (list of) bioconductor assay containers into a FacileDataSet. — as.FacileDataSet" />
<meta property="og:description" content="This function assumes you are only extracting one assay from the assay
container and creating a FacileDataSet from it. If the bioc-containers
you are using can support more than one assay, specify which one to extract
using the source_assay parameter, otherwise the first assay in the
container will be taken. The source_assay you are using to initialize the
FacileDataSet here will become its default_assay when its used later.
The feature_info table is populated by the particular fData (mcols,
y$genes, etc.) from the assay container. The pData from the assay
container will be ingested as well, so do your best to ensure that this
represents the most complete version of the pData for the FacileDataSet
you will be creating." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FacileData</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.12.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/FacileData.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/FacileDataSet-architecture.html">FacileDataSet Architecture</a>
    </li>
    <li>
      <a href="../articles/FacileDataSet-assembly.html">FacileDataSet Assembly</a>
    </li>
    <li>
      <a href="../articles/FacileDataSet.html">FacileDataSet Overview</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/facilebio/FacileData/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Converts a (list of) bioconductor assay containers into a FacileDataSet.</h1>
    <small class="dont-index">Source: <a href='https://github.com/facilebio/FacileData/blob/master/R/as.FacileDataSet.R'><code>R/as.FacileDataSet.R</code></a></small>
    <div class="hidden name"><code>as.FacileDataSet.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function assumes you are only extracting one assay from the assay
container and creating a <code>FacileDataSet</code> from it. If the bioc-containers
you are using can support more than one assay, specify which one to extract
using the <code>source_assay</code> parameter, otherwise the first assay in the
container will be taken. The <code>source_assay</code> you are using to initialize the
<code>FacileDataSet</code> here will become its <code>default_assay</code> when its used later.</p>
<p>The <code>feature_info</code> table is populated by the particular <code>fData</code> (<code>mcols</code>,
<code>y$genes</code>, etc.) from the assay container. The <code>pData</code> from the assay
container will be ingested as well, so do your best to ensure that this
represents the most complete version of the pData for the FacileDataSet
you will be creating.</p>
    </div>

    <pre class="usage"><span class='fu'>as.FacileDataSet</span><span class='op'>(</span>
  <span class='va'>x</span>,
  <span class='va'>path</span>,
  <span class='va'>assay_name</span>,
  <span class='va'>assay_type</span>,
  <span class='va'>source_assay</span>,
  assay_description <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span><span class='st'>"Description for "</span>, <span class='va'>assay_name</span><span class='op'>)</span>,
  dataset_name <span class='op'>=</span> <span class='st'>"DEFAULT_NAME"</span>,
  dataset_meta <span class='op'>=</span> <span class='cn'>NULL</span>,
  organism <span class='op'>=</span> <span class='st'>"unspecified"</span>,
  page_size <span class='op'>=</span> <span class='fl'>2</span><span class='op'>^</span><span class='fl'>12</span>,
  cache_size <span class='op'>=</span> <span class='fl'>2e+05</span>,
  chunk_rows <span class='op'>=</span> <span class='fl'>5000</span>,
  chunk_cols <span class='op'>=</span> <span class='st'>"ncol"</span>,
  chunk_compression <span class='op'>=</span> <span class='fl'>5</span>,
  covariate_def <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S3 method for list</span>
<span class='fu'>as.FacileDataSet</span><span class='op'>(</span>
  <span class='va'>x</span>,
  <span class='va'>path</span>,
  <span class='va'>assay_name</span>,
  <span class='va'>assay_type</span>,
  source_assay <span class='op'>=</span> <span class='cn'>NULL</span>,
  assay_description <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span><span class='st'>"Description for "</span>, <span class='va'>assay_name</span><span class='op'>)</span>,
  dataset_name <span class='op'>=</span> <span class='st'>"DEFAULT_NAME"</span>,
  dataset_meta <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='op'>)</span>,
  organism <span class='op'>=</span> <span class='st'>"unspecified"</span>,
  prune_dataset_meta <span class='op'>=</span> <span class='cn'>TRUE</span>,
  page_size <span class='op'>=</span> <span class='fl'>2</span><span class='op'>^</span><span class='fl'>12</span>,
  cache_size <span class='op'>=</span> <span class='fl'>2e+05</span>,
  chunk_rows <span class='op'>=</span> <span class='fl'>5000</span>,
  chunk_cols <span class='op'>=</span> <span class='st'>"ncol"</span>,
  chunk_compression <span class='op'>=</span> <span class='fl'>5</span>,
  covariate_def <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>The bioconductor assay container to extract data from</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>the directory to create the faciledataset into. Will create
a default directory in the current working directory if not specified.
This directory should not yet exist.</p></td>
    </tr>
    <tr>
      <th>assay_name</th>
      <td><p>The name to use when storing the assay matrix from
<code>x</code> into the faciledataset.</p></td>
    </tr>
    <tr>
      <th>assay_type</th>
      <td><p>what type of assay is this? rnaseq, microarray, nanostring,
isoseq (isoform expression), etc.</p></td>
    </tr>
    <tr>
      <th>source_assay</th>
      <td><p>the name of the assay element in <code>x</code> to extract
for use.</p></td>
    </tr>
    <tr>
      <th>dataset_name</th>
      <td><p>the <code>name</code> attribute of the FacileDataSet <code>meta.yaml</code>
file.</p></td>
    </tr>
    <tr>
      <th>dataset_meta</th>
      <td><p>a named (by names(x)) with meta data about the datasets
that appear in the list of datasets <code>x</code>. List elements per dataset should
minimally include a <code>description</code> and <code>url</code> string.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p>This is used to fetch the appropriate genesets when this
dataset is used with the facileexplorer. Put species name here, ie.
<code>"Homo sapiens"</code>, <code>"Mus musculus"</code>, etc. Default is <code>"unspecified"</code>,
which isn't really helpful.</p></td>
    </tr>
    <tr>
      <th>page_size</th>
      <td><p>parameter to tweak SQLite</p></td>
    </tr>
    <tr>
      <th>cache_size</th>
      <td><p>parameter to tweak SQLite</p></td>
    </tr>
    <tr>
      <th>chunk_rows</th>
      <td><p>parameter to tweak HDF5</p></td>
    </tr>
    <tr>
      <th>chunk_cols</th>
      <td><p>parameter to tweak HDF5</p></td>
    </tr>
    <tr>
      <th>chunk_compression</th>
      <td><p>parameter to tweak HDF5</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>more args</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code><a href='FacileDataSet.html'>FacileDataSet()</a></code></p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>A <code>FacileDataSet</code> can be created from a number of different Bioconductor
containers, such as a <code><a href='https://rdrr.io/pkg/Biobase/man/class.ExpressionSet.html'>Biobase::ExpressionSet</a></code>,
<code><a href='https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html'>SummarizedExperiment::SummarizedExperiment</a></code>, or an <code><a href='https://rdrr.io/pkg/edgeR/man/DGEList.html'>edgeR::DGEList</a></code>. To
create a <code>FacileDataSet</code> that spans multiple Bioc containters, i.e. you may
have one ExpressionSet per indication in the TCGA. You can make
<code>FacileDataSet</code> to encompass the data from <em>all</em> of these indications by
providing a <code>list</code> of <code>ExpressionSet</code>s. The <code>list</code> should have its <code><a href='https://rdrr.io/r/base/names.html'>names()</a></code>
set to each of the TCGA indications ("BLCA", "BRCA", etc.) the data came
from.</p>
    <h2 class="hasAnchor" id="sample-covariates"><a class="anchor" href="#sample-covariates"></a>Sample Covariates</h2>

    


<p>The <code>pData</code> data.frame object will be picked off from all of the containers
provided in the list of datasets you are using to create the FacileDataSet.
<code>dataset</code> and <code>sample_id</code> columns will be forcibly added (or modified) as
columns to all of the individual <code>pData</code> data.frames.</p>
<p>In order to insert the entirety of the <code>pData</code> elements into the internal
<code>sample_covariate</code> table, we rely on the <code><a href='https://dplyr.tidyverse.org/reference/bind.html'>dplyr::bind_rows</a></code> function to
create an uber <code>data.frame</code> which will be converted into an
entity-attribute-value table. Note that when row-binding, columns are matched
by name, and any missing columns with be filled with <code>NA</code>.</p>
<p><code>ExpressionSet</code> pData <code>data.frames</code> should have an attribute called 'label', which
will be a named character vector with a description for each column. In the case of
a <code>SummarizedExperiment</code>, the <code>colData</code> should have named list in the <code>metadata</code>
slot with a character description of each column.</p>
<p><code>ExpressionSet</code>s should have a short textual description of the facet/dataset in
the <code>annotation</code> slot. Similarly, <code>SummarizedExperiment</code>s should have a list
in the <code>metadata</code> slot with <code>url</code> and <code>description</code> for the facet/dataset.</p>
<p>Please ensure that the covariates across the <code>pData</code> data.frames have already
been harmonized!</p>
    <h2 class="hasAnchor" id="feature-meta-information"><a class="anchor" href="#feature-meta-information"></a>Feature meta-information</h2>

    


<p>The feature information (aka "fData") are stored in an internal
<code>feature_info</code> SQLite table within the <code>FacileDataSet</code>. The information to
populate this table will be retrieved from the corresponding <code>fData</code>-like
<code>data.frame</code> from <strong>the first</strong> given bioc-container in the list.
This <code>data.frame</code> must define the following columns:</p><ul>
<li><p>"feature_type": <code>string</code>, one of: <code>"entrez"</code>, <code>"ensgid"</code>, <code>"enstid"</code>,
<code>"genomic"</code>, <code>"custom"</code>.</p></li>
<li><p>"feature_id": <code>string</code></p></li>
<li><p>"name": <code>string</code></p></li>
<li><p>"meta": <code>string</code></p></li>
<li><p>"effective_length": <code>integer</code></p></li>
<li><p>"source": <code>string</code></p></li>
</ul>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Steve Lianoglou, Vincent Rouilly, Peter Haverty, Denali Therapeutics, Genentech.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


