<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ FacileData</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FacileData</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.12.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/FacileData.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/FacileDataSet-architecture.html">FacileDataSet Architecture</a>
    </li>
    <li>
      <a href="../articles/FacileDataSet-assembly.html">FacileDataSet Assembly</a>
    </li>
    <li>
      <a href="../articles/FacileDataSet.html">FacileDataSet Overview</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/facilebio/FacileData/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="FacileData-package.html">FacileData-package</a></code> </p>
        </td>
        <td><p>FacileData: A fluent API for accessing multi-assay high-throughput genomics data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="FacileDataSet.html">FacileDataSet()</a></code> </p>
        </td>
        <td><p>Instantiates a FacileDataSet object from disk.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="adata.html">adata()</a></code> </p>
        </td>
        <td><p>Bioc-container specific assay data extraction functions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addFacileAssaySet.html">addFacileAssaySet()</a></code> </p>
        </td>
        <td><p>Adds new set of assay data for all samples in a FacileDataSet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="append_facile_feature_info.html">append_facile_feature_info()</a></code> </p>
        </td>
        <td><p>Appends new features to <code>feature_info</code> table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="append_facile_table.html">append_facile_table()</a></code> </p>
        </td>
        <td><p>Adds rows to a table in a FacileDataSet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="as.BiocContainer.html">as.DGEList()</a></code> <code><a href="as.BiocContainer.html">as.ExpressionSet()</a></code> <code><a href="as.BiocContainer.html">as.SummarizedExperiment()</a></code> </p>
        </td>
        <td><p>Converts a "facile result" to a traditional Bioconductor assay container.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="as.EAVtable.html">as.EAVtable()</a></code> </p>
        </td>
        <td><p>Convert a <code>pData</code> data.frame to a melted EAV table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="as.FacileDataSet.html">as.FacileDataSet()</a></code> </p>
        </td>
        <td><p>Converts a (list of) bioconductor assay containers into a FacileDataSet.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="as_facile_frame.html">as_facile_frame()</a></code> </p>
        </td>
        <td><p>Converts a normal tibble/data.frame to a facile_frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assay_feature_info.FacileDataSet.html">assay_feature_info(<i>&lt;FacileDataSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>Materializes a table with all feature information for a given assay.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assay_feature_type.html">assay_feature_type()</a></code> </p>
        </td>
        <td><p>Returns the feature_type for a given assay</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assay_info.html">assay_info()</a></code> </p>
        </td>
        <td><p>Fetches assay meta information for the assays stored in a FacileDataStore</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assay_info_over_samples.html">assay_info_over_samples()</a></code> </p>
        </td>
        <td><p>Identify the number of each assay run across specific samples.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assay_sample_info.html">assay_sample_info()</a></code> </p>
        </td>
        <td><p>Utility functions to get row and column indices of rnaseq hdf5 files.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assay_units.html">assay_units()</a></code> </p>
        </td>
        <td><p>Units of measure in an assay</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assemble_example_dataset.html">assemble_example_dataset()</a></code> </p>
        </td>
        <td><p>Assembles an example facile dataset to play with</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertions.html">assert_sample_subset()</a></code> <code><a href="assertions.html">check_sample_subset()</a></code> <code><a href="assertions.html">test_sample_subset()</a></code> <code><a href="assertions.html">assert_facet_descriptor()</a></code> <code><a href="assertions.html">is_facet_descriptor()</a></code> <code><a href="assertions.html">assert_assay_feature_descriptor()</a></code> <code><a href="assertions.html">is_assay_feature_descriptor()</a></code> <code><a href="assertions.html">assert_expression_result()</a></code> <code><a href="assertions.html">is_expression_result()</a></code> <code><a href="assertions.html">assert_sample_statistics()</a></code> <code><a href="assertions.html">is_sample_statistics()</a></code> <code><a href="assertions.html">assert_sample_covariates()</a></code> <code><a href="assertions.html">is_sample_covariates()</a></code> <code><a href="assertions.html">assert_columns()</a></code> <code><a href="assertions.html">has_columns()</a></code> <code><a href="assertions.html">assert_covariate_definitions()</a></code> <code><a href="assertions.html">is_covariate_definitions()</a></code> </p>
        </td>
        <td><p>Check to see that samples are referenced correctly</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="biocbox.html">biocbox()</a></code> </p>
        </td>
        <td><p>Materialize a Bioconductor assay container from some facile object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cSurv.html">as_cSurv()</a></code> <code><a href="cSurv.html">as_Surv()</a></code> </p>
        </td>
        <td><p>cSurv is a character representation of survival::surv ###</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cast_covariate.html">cast_covariate()</a></code> </p>
        </td>
        <td><p>Casts the character EAV values to their R-native defined types.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_categorical.html">check_categorical()</a></code> <code><a href="check_categorical.html">assert_categorical()</a></code> <code><a href="check_categorical.html">test_categorical()</a></code> </p>
        </td>
        <td><p>Check to see if a vector is categorical (character or string)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_facile_data_set.html">check_facile_data_set()</a></code> <code><a href="check_facile_data_set.html">assert_facile_data_set()</a></code> <code><a href="check_facile_data_set.html">test_facile_data_set()</a></code> </p>
        </td>
        <td><p>Check if argument is a FacileDataSet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_facile_data_store.html">check_facile_data_store()</a></code> <code><a href="check_facile_data_store.html">assert_facile_data_store()</a></code> <code><a href="check_facile_data_store.html">test_facile_data_store()</a></code> </p>
        </td>
        <td><p>Check if argument is a FacileDataStore</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="conform_data_frame.html">conform_data_frame()</a></code> </p>
        </td>
        <td><p>Arranges the columns of one data.frame to another</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="covariate_definitions.html">covariate_definitions()</a></code> </p>
        </td>
        <td><p>Get description of sample metadata columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="covariate_meta_info.html">covariate_meta_info()</a></code> </p>
        </td>
        <td><p>Retrieve the meta information about a covariate for EAV decoding</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="create_assay_feature_descriptor.html">create_assay_feature_descriptor()</a></code> </p>
        </td>
        <td><p>Creates a feature descriptor for interactive ease</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dbfn.html">dbfn()</a></code> </p>
        </td>
        <td><p>Get location of the FacileDataSet database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dot-level_biotypes.html">.level_biotypes()</a></code> </p>
        </td>
        <td><p>Utility function to "factorize" biotypes into an order we care about.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ds_annot.html">ds_annot()</a></code> </p>
        </td>
        <td><p>Bioc-container specific data set annotation extraction functions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="eav-metadata.html">eav_metadata_create()</a></code> </p>
        </td>
        <td><p>Create a facile covariate definition file from a sample <code>pData</code> data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="eav-right-censor.html">eav_encode_right_censored()</a></code> <code><a href="eav-right-censor.html">eav_decode_right_censored()</a></code> </p>
        </td>
        <td><p>Entity-attribute-value encodings for survival data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="eav_encode.html">eav_encode()</a></code> </p>
        </td>
        <td><p>Encodes column(s) from <code>pData</code> into character values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="eav_encode_covariate.html">eav_encode_covariate()</a></code> </p>
        </td>
        <td><p>Encodes column(s) from <code>pData</code> into character values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="eav_metadata_merge.html">eav_metadata_merge()</a></code> </p>
        </td>
        <td><p>Merge inferred and explicit covariate column metadata.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="eavdef_for_column.html">eavdef_for_column()</a></code> </p>
        </td>
        <td><p>Generate entity-attribute-value definition for a column in a data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="exampleFacileDataSet.html">exampleFacileDataSet()</a></code> </p>
        </td>
        <td><p>Retrieves an example FacileDataSet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="executeSQL.html">executeSQL()</a></code> </p>
        </td>
        <td><p>Execute multiple queries against a database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_transcribed_info_from_ensembl_gtf.html">extract_transcribed_info_from_ensembl_gtf()</a></code> </p>
        </td>
        <td><p>Extract gene- and transcript-level information from an ENSEMBL gtf.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="facet_frame.html">facet_frame()</a></code> </p>
        </td>
        <td><p>Retrieves grouping table for samples within a FacileDataSet.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="facilitate.html">facilitate()</a></code> </p>
        </td>
        <td><p>Converts an arbitrary object into one that works in the facile ecosystem.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fdata.html">fdata()</a></code> </p>
        </td>
        <td><p>BioC-container specific fData extraction functions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fds.html">fds()</a></code> <code><a href="fds.html">`fds&lt;-`()</a></code> <code><a href="fds.html">set_fds()</a></code> </p>
        </td>
        <td><p>Get or set the FacileDataStore for an object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="feature_name_map.html">assay_feature_name_map()</a></code> <code><a href="feature_name_map.html">feature_name_map()</a></code> </p>
        </td>
        <td><p>Returns table of names and aliases for features.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="feature_types.html">feature_types()</a></code> </p>
        </td>
        <td><p>Enumerate the types of feature stored in a FacileDataSet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="features.html">features()</a></code> </p>
        </td>
        <td><p>Returns a table of information about the features (from an assay, or ...)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetch_assay_data.html">fetch_assay_data()</a></code> <code><a href="fetch_assay_data.html">with_assay_data(<i>&lt;facile_frame&gt;</i>)</a></code> </p>
        </td>
        <td><p>Fetch assay data from single assay of choice</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetch_assay_score.FacileDataSet.html">fetch_assay_score(<i>&lt;FacileDataSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>Helper function to get sample assay data from single or aggregate features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetch_assay_score.html">fetch_assay_score()</a></code> </p>
        </td>
        <td><p>NOTE: is fetch_assay_score really necessary?</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetch_custom_sample_covariates.FacileDataSet.html">fetch_custom_sample_covariates(<i>&lt;FacileDataSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>Fetches custom (user) annotations for a given user prefix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetch_sample_statistics.FacileDataSet.html">fetch_sample_statistics(<i>&lt;FacileDataSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>Fetch the sample statistics for sets of samples in the warehouse</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetch_sample_statistics.html">fetch_sample_statistics()</a></code> </p>
        </td>
        <td><p>NOTE: fetch_sample_statistics -&gt; <code>fetch_assay_covariates</code>
Issue #2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetch_sample_statistics.default.html">fetch_sample_statistics(<i>&lt;default&gt;</i>)</a></code> </p>
        </td>
        <td><p>Issue #2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fetch_samples.FacileDataSet.html">fetch_samples(<i>&lt;FacileDataSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>Fetches a sample descriptor that matches the filter criterion</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_features.FacileDataSet.html">filter_features(<i>&lt;FacileDataSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>Filter against the sample_covariate_tbl as if it were wide.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_samples.FacileDataSet.html">filter_samples(<i>&lt;FacileDataSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>Filter against the sample_covariate_tbl EAV table as if it were wide.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flog.html">flog()</a></code> </p>
        </td>
        <td><p>Generates a logging message using glue and crayon, with some bells/whistles.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flog_level.html">flog_level()</a></code> </p>
        </td>
        <td><p>Retrieves the currently set logging level</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="freplace_na.html">freplace_na()</a></code> </p>
        </td>
        <td><p>Replaces NA's with specified values.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_info_tbl.html">gene_info_tbl()</a></code> </p>
        </td>
        <td><p>Mimics the old <code>gene_info</code> table.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="has_feature_type.html">has_feature_type()</a></code> </p>
        </td>
        <td><p>Test if a given feature type is stored in a FacileDataSet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hdf5fn.html">hdf5fn()</a></code> </p>
        </td>
        <td><p>Get location of the FacileDataSet HDF5 file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="infer_feature_type.html">infer_feature_type()</a></code> </p>
        </td>
        <td><p>Guesses the type of feature identifiers from a character vector.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="initializeFacileDataSet.html">initializeFacileDataSet()</a></code> </p>
        </td>
        <td><p>Create an empty FacileDataSet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="is.FacileDataSet.html">is.FacileDataSet()</a></code> </p>
        </td>
        <td><p>Class and validity checker for FacileDataSet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="join_samples.html">join_samples()</a></code> </p>
        </td>
        <td><p>Filters the samples down in a dataset to ones specified</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="labeled.html">name()</a></code> </p>
        </td>
        <td><p>Labeled acts like interface to reactive modules.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="meta-info.html">meta_file()</a></code> <code><a href="meta-info.html">meta_info()</a></code> <code><a href="meta-info.html">default_assay(<i>&lt;FacileDataSet&gt;</i>)</a></code> <code><a href="meta-info.html">dataset_definitions()</a></code> </p>
        </td>
        <td><p>Path to the meta information YAML file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="nameit.html">nameit()</a></code> </p>
        </td>
        <td><p>Ensures that a vector has names for all elements if it has names for any</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="normalize_assay_data.html">normalize_assay_data()</a></code> </p>
        </td>
        <td><p>Helper functions to normalize assay data into log2 space.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="organism.FacileDataSet.html">organism(<i>&lt;FacileDataSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>Retrieves the organism the data is defined over</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_sample_criterion.html">parse_sample_criterion()</a></code> </p>
        </td>
        <td><p>Creates a filter expression to select samples based on value of a covariate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pdata.html">pdata()</a></code> </p>
        </td>
        <td><p>Bioc-container specific pData extraction functions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pdata_metadata.html">pdata_metadata()</a></code> </p>
        </td>
        <td><p>Bioc-container specific pData extraction functions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="primary_key.html">primary_key()</a></code> </p>
        </td>
        <td><p>Query a table to identify its primary key(s)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="remove_batch_effect.html">remove_batch_effect()</a></code> </p>
        </td>
        <td><p>Regress out confounding variables from a data matrix.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sample-covariates.html">fetch_sample_covariates()</a></code> <code><a href="sample-covariates.html">with_sample_covariates()</a></code> </p>
        </td>
        <td><p>Appends covariate columns to a query result</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sample_stats_tbl.html">sample_stats_tbl()</a></code> </p>
        </td>
        <td><p>Mimics old sample_stats table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="samples.FacileDataSet.html">samples(<i>&lt;FacileDataSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>Retrieves the sample identifiers for all samples in a FacileDataSet.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="save_custom_sample_covariates.html">save_custom_sample_covariates()</a></code> </p>
        </td>
        <td><p>Saves custom sample covariates to a FacileDataSet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="set_class.html">set_class()</a></code> </p>
        </td>
        <td><p>Set the class of an object and return the object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="simple-eav-decode-functions.html">eav_decode_real()</a></code> <code><a href="simple-eav-decode-functions.html">eav_encode_real()</a></code> <code><a href="simple-eav-decode-functions.html">eav_encode_logical()</a></code> <code><a href="simple-eav-decode-functions.html">eav_decode_logical()</a></code> <code><a href="simple-eav-decode-functions.html">eav_encode_cSurv()</a></code> <code><a href="simple-eav-decode-functions.html">eav_decode_cSurv()</a></code> <code><a href="simple-eav-decode-functions.html">eav_decode_categorical()</a></code> <code><a href="simple-eav-decode-functions.html">eav_encode_categorical()</a></code> </p>
        </td>
        <td><p>Entity-attribute-value decoding for real values.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spread_assay_data.html">spread_assay_data()</a></code> </p>
        </td>
        <td><p>Takes a result from fetch_expression and spreads out genes across columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spread_covariates.html">spread_covariates()</a></code> </p>
        </td>
        <td><p>Spreads the covariates returned from database into wide data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sqlFromFile.html">sqlFromFile()</a></code> </p>
        </td>
        <td><p>Utility function to send more than one sql command to the database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summary.eav_covariates.html">summary(<i>&lt;eav_covariates&gt;</i>)</a></code> </p>
        </td>
        <td><p>Provides a summary table of sample covariates.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test-helpers.html">example_sample_covariates()</a></code> <code><a href="test-helpers.html">example_meta()</a></code> <code><a href="test-helpers.html">example_sample_covariate_definitions()</a></code> </p>
        </td>
        <td><p>Fetches exemplar data for unit testing</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="validate_covariate_def_list.html">validate_covariate_def_list()</a></code> </p>
        </td>
        <td><p>Validates that a covariate defintion list reasonably describes a data.frame.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="with_feature_info.html">with_feature_info()</a></code> </p>
        </td>
        <td><p>Append feature information columns to (feature-rows)</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Steve Lianoglou, Vincent Rouilly, Peter Haverty, Denali Therapeutics, Genentech.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


