<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>FacileDataSet Assembly • FacileData</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="FacileDataSet Assembly">
<meta property="og:description" content="FacileData">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FacileData</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.98.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/FacileData.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/FacileDataSet-architecture.html">FacileDataSet Architecture</a>
    </li>
    <li>
      <a href="../articles/FacileDataSet-assembly.html">FacileDataSet Assembly</a>
    </li>
    <li>
      <a href="../articles/FacileDataSet.html">FacileDataSet Overview</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/facilebio/FacileData/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="FacileDataSet-assembly_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>FacileDataSet Assembly</h1>
                        <h4 class="author">Steve Lianoglou</h4>
            
            <h4 class="date">5/17/2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/facilebio/FacileData/blob/master/vignettes/FacileDataSet-assembly.Rmd"><code>vignettes/FacileDataSet-assembly.Rmd</code></a></small>
      <div class="hidden name"><code>FacileDataSet-assembly.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>The <code>FacileDataSet</code> is a reference implementation of a multi-assay datastore that implements the <em>FacileData</em> API. It uses a SQLite database to store feature- and sample-level metadata, and an HDF5 file to store any number of assays over its samples. Both of these technologies allow quick and efficient access to arbitrary subsets of data stored on disk without having to load all of the data into RAM, making this datastore ideal for housing data like <em>the entirety</em> of the data from <a href="https://cancergenome.nih.gov/">The Cancer Genome Atlas</a>.</p>
<p>The <a href="http://github.com/facileverse/FacileTcgaDataSet">FacileTcgaDataSet</a> package outlines how to assemble the multiple assays from the TCGA (RNA-seq, miRNA-seq, CNV, etc.) into one FacileDataSet, and provides examples of how to access these data using the FacileData API.</p>
<p>We’ll be a bit less ambitious in this vignette, and only assemble two different RNA-seq datasets into one FacileDataSet. This won’t exercise our ability to query across different assays, but we will use this dataset for downstream examples here, and in vignettes in other facile packages.</p>
<p>We will use this vignette to assemble these data in a stepwise fashion, highlighting some best practices for doing so. These maneuvers will also be encapsulated in the <code><a href="../reference/assemble_example_dataset.html">assemble_example_dataset()</a></code> in this package, for other resources to use, as well.</p>
<div class="note">
<p>We will be working on a <a href="https://github.com/facileverse/FacileBioc">FacileBioc</a> package that will provide the FacileData API directly over native Bioconductor assay containers. This will allow users to work with these objects directly with out explicitly converting them into a FacileDataSet first. Until then, follow the instructions here to convert data you already have into a FacileDataSet.</p>
</div>
</div>
<div id="data-set-curation" class="section level2">
<h2 class="hasAnchor">
<a href="#data-set-curation" class="anchor"></a>Data Set Curation</h2>
<p>The easiest way to create a <code>FacileDataSet</code> is to pass a <em>named</em> list of <a href="#well-groomed-datasets">well-groomed</a> bioconductor assay containers, like a <code>SummarizedExperiment</code> or <code>DGEList</code> (or <code>ExpressionSet</code>) to the <code>as.FacileDataSet</code> function.</p>
<p>A <code>FacileDataSet</code> is stored on disk as a well-structured directory. You can look at the <a href="#faciledataset-architecture">FacileDataSet Architecture</a> section for the full details, but the important information to know is that:</p>
<ol style="list-style-type: decimal">
<li>The dense assay matrices of the <code>SummarizedExperiment</code> are stored in an HDF5 file.</li>
<li>Feature (row) and sample (column) metadata (<code>fData</code> and <code>pData</code>, respectively) are stored in an SQLite database.</li>
<li>The <code>pData</code>/<code>colData</code> data.frames of each <code>SummarizedExperiment</code> are stored in an SQLite table in an entity-attribute-value-table (EAV) format in order to support covariates that are both common to all <code>SummarizedExperiment</code>s and unique to others.</li>
<li>There is a <code>meta.yaml</code> file that stores important information about the dataset itself. Perhaps most important is the <code>sample_covariates</code> section that dictates how each covariate stored in the EAV table is decoded back into its native R-representation. More details are provided in the <a href="#entity-attribute-value-table">Entity-Attribute-Value Table</a> section.</li>
</ol>
<p>Here we will use two already-assembled (but unrelated) RNA-seq datasets easily available within the bioconductor ecosystem for data assembly:</p>
<ol style="list-style-type: decimal">
<li><p>The <a href="http://bioconductor.org/packages/airway">airway</a> dataset. This package provides a RangedSummarizedExperiment object of read counts in genes for an RNA-Seq experiment on four human airway smooth muscle cell lines treated with dexamethasone.</p></li>
<li><p>The <a href="http://bioconductor.org/packages/parathyroidSE">parathyroidSE</a> dataset. This package provides RangedSummarizedExperiment objects of read counts in genes and exonic parts for paired-end RNA-Seq data from experiments on primary cultures of parathyroid tumors.</p></li>
</ol>
<div class="tip">
<p>If you are looking for help wrangling raw RNA-seq data into something like a SummarizedExperiment object, we currently favor processing the reads using either <a href="https://combine-lab.github.io/salmon/">salmon</a> or <a href="https://pachterlab.github.io/kallisto/">kallisto</a>, then importing these results using the <a href="http://bioconductor.org/packages/tximport">tximport</a> package.</p>
</div>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a></span><span class="op">)</span>
<span class="co"># Note that you will need the airway and parathyroidSE packages installed.</span>
<span class="co"># If you don't have those, install them like so:</span>
<span class="co"># </span>
<span class="co"># BiocManager::install(c("airway", "parathyroidSE"))</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"airway"</span>, package <span class="op">=</span> <span class="st">"airway"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"parathyroidGenesSE"</span>, package <span class="op">=</span> <span class="st">"parathyroidSE"</span><span class="op">)</span>

<span class="co"># We load this after the Bioconductor stuff, because there are some facile</span>
<span class="co"># generic functions that collide w/ Bioconductor ones. `samples` is defined</span>
<span class="co"># in Biocbase, for instance.</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/facilebio/FacileData">FacileData</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="well-groomed-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#well-groomed-datasets" class="anchor"></a>Well Groomed Datasets</h2>
<p>Keep in mind that we will be assembling these different datasets into one container that will combine the assay data from both experiments into one, as well as their gene- and sample- level metadata.</p>
<div id="sample-data-harmonization" class="section level3">
<h3 class="hasAnchor">
<a href="#sample-data-harmonization" class="anchor"></a>Sample Data Harmonization</h3>
<p>An often under-valued part of the high-througput experiments we tend to generate is how we document the phenotypic variables over our samples, ie. the things that populate the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData()</a></code> of a SummarizedExperiment. This part is particularly important when assembling different datasets into one.</p>
<p>Put simply, we will want to ensure that where there is overlap in the “pnenotypic space” between different SummarizedExperiments, their variable (column) names are the same, and their encodings (values in the columns) are the same when they are talking about the same thing. We’ll call this process <em>“harmonization”</em>.</p>
<p>As an example, one batch of experiments might have a <code>"sex"</code> covariate, encoded with <code>"male"</code> and <code>"female"</code> entries. Another might encode <code>"sex"</code> using <code>"m"</code> and <code>"f"</code> values. Yet another might have labelled <code>"sex"</code> as <code>"gender"</code>, and encoded the values as <code>"MALE"</code> and <code>"FEMALE"</code>. When assembling these different experiment togheter, you want to ensure that the <code>"sex"</code> covariate is named the same across the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData()</a></code> objects of the experiments you are assembling, and further that all of its entires are encoded uniformly as <code>"male"</code> and <code>"female"</code>.</p>
<div class="note">
<p>There is no requirement in this sample-data harmonization process to ensure that every experiment has <em>the same</em> covariates. Some experiments do not track the same phenotypic variables as others. This is fine: every SummarizedExperiment does not need the same elements in its <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData()</a></code>. We just want to ensure that when there is overlap, the column names and encodings match across datasets.</p>
</div>
<p>Another thing to consider when encoding covariate values is to ensure that they are, themselves, <strong>encoded as valid R variable names</strong>. For instance, suppose we have a <code>time</code> factor, with levels like <code>"24h"</code>, and <code>"48h"</code>, these should be re-encoded to something like <code>"h24"</code> and <code>"h48"</code>. Here’s why: often times while you’re running different sorts of exploratory data analyses, through some sequence of <code>"gather"</code>-ing, <code>"spread"</code>-ing, and joining sample-level data, it is not uncommon for factor levels to end up as columns of a data.frame. Ensuring that the factor loevels are “proper R variable” names up front avoids problems you might run into when using non standard evaluation with whatever dplyr mojo, <code><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix(...)</a></code>, or whatever else these data often find themselves subject to.</p>
<div id="harmonization-put-to-practice" class="section level4">
<h4 class="hasAnchor">
<a href="#harmonization-put-to-practice" class="anchor"></a>Harmonization put to practice</h4>
<p>Let’s take a look at the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData()</a></code> from our two datasets, and see how we might harmonize them:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##            SampleName    cell   dex albut        Run avgLength Experiment
## SRR1039508 GSM1275862  N61311 untrt untrt SRR1039508       126  SRX384345
## SRR1039509 GSM1275863  N61311   trt untrt SRR1039509       126  SRX384346
## SRR1039512 GSM1275866 N052611 untrt untrt SRR1039512       126  SRX384349
## SRR1039513 GSM1275867 N052611   trt untrt SRR1039513        87  SRX384350
## SRR1039516 GSM1275870 N080611 untrt untrt SRR1039516       120  SRX384353
## SRR1039517 GSM1275871 N080611   trt untrt SRR1039517       126  SRX384354
## SRR1039520 GSM1275874 N061011 untrt untrt SRR1039520       101  SRX384357
## SRR1039521 GSM1275875 N061011   trt untrt SRR1039521        98  SRX384358
##               Sample    BioSample
## SRR1039508 SRS508568 SAMN02422669
## SRR1039509 SRS508567 SAMN02422675
## SRR1039512 SRS508571 SAMN02422678
## SRR1039513 SRS508572 SAMN02422670
## SRR1039516 SRS508575 SAMN02422682
## SRR1039517 SRS508576 SAMN02422673
## SRR1039520 SRS508579 SAMN02422683
## SRR1039521 SRS508580 SAMN02422677</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">parathyroidGenesSE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##          run experiment patient treatment time submission     study    sample
## 1  SRR479052  SRX140503       1   Control  24h  SRA051611 SRP012167 SRS308865
## 2  SRR479053  SRX140504       1   Control  48h  SRA051611 SRP012167 SRS308866
## 3  SRR479054  SRX140505       1       DPN  24h  SRA051611 SRP012167 SRS308867
## 4  SRR479055  SRX140506       1       DPN  48h  SRA051611 SRP012167 SRS308868
## 5  SRR479056  SRX140507       1       OHT  24h  SRA051611 SRP012167 SRS308869
## 6  SRR479057  SRX140508       1       OHT  48h  SRA051611 SRP012167 SRS308870
## 7  SRR479058  SRX140509       2   Control  24h  SRA051611 SRP012167 SRS308871
## 8  SRR479059  SRX140510       2   Control  48h  SRA051611 SRP012167 SRS308872
## 9  SRR479060  SRX140511       2       DPN  24h  SRA051611 SRP012167 SRS308873
## 10 SRR479061  SRX140511       2       DPN  24h  SRA051611 SRP012167 SRS308873
## 11 SRR479062  SRX140512       2       DPN  48h  SRA051611 SRP012167 SRS308874
## 12 SRR479063  SRX140513       2       OHT  24h  SRA051611 SRP012167 SRS308875
## 13 SRR479064  SRX140513       2       OHT  24h  SRA051611 SRP012167 SRS308875
## 14 SRR479065  SRX140514       2       OHT  48h  SRA051611 SRP012167 SRS308876
## 15 SRR479066  SRX140515       3   Control  24h  SRA051611 SRP012167 SRS308877
## 16 SRR479067  SRX140516       3   Control  48h  SRA051611 SRP012167 SRS308878
## 17 SRR479068  SRX140517       3       DPN  24h  SRA051611 SRP012167 SRS308879
## 18 SRR479069  SRX140518       3       DPN  48h  SRA051611 SRP012167 SRS308880
## 19 SRR479070  SRX140519       3       OHT  24h  SRA051611 SRP012167 SRS308881
## 20 SRR479071  SRX140520       3       OHT  48h  SRA051611 SRP012167 SRS308882
## 21 SRR479072  SRX140521       4   Control  48h  SRA051611 SRP012167 SRS308883
## 22 SRR479073  SRX140522       4       DPN  24h  SRA051611 SRP012167 SRS308884
## 23 SRR479074  SRX140523       4       DPN  48h  SRA051611 SRP012167 SRS308885
## 24 SRR479075  SRX140523       4       DPN  48h  SRA051611 SRP012167 SRS308885
## 25 SRR479076  SRX140524       4       OHT  24h  SRA051611 SRP012167 SRS308886
## 26 SRR479077  SRX140525       4       OHT  48h  SRA051611 SRP012167 SRS308887
## 27 SRR479078  SRX140525       4       OHT  48h  SRA051611 SRP012167 SRS308887</code></pre>
<p>Here are some things to consider:</p>
<ol style="list-style-type: decimal">
<li><p>There are only a subset of columns from the two experiments that tell us something about the characteristics of their samples, such as the “cell”, and “dex” columns of the airway dataset, and patient, treatment, and time, column of the parathyroid dataset. The “albut” column in the airway dataset is non-informative in this subset of the data, since it is always the same value, so we’ll drop that, too.</p></li>
<li><p>When combining these two experiments, we can probably consider the “dex” column in the airway dataset as a “treatment”, so we can combine them with the same column in the parathyroid dataset. Further, the encodings of the dex column might be transformed from “untrt” and “trt” to “control” and “dex”.</p></li>
<li><p>The “time” column in the parathyroid experiment should be translated into valid R variable names, so we’ll convert <code>24h</code> to <code>hrs24</code>. The entries in the parathyroid “patient” column will also be munged in a similar fashion. We’ve found naming covariates of this type as “subject_id” covers many cases when working across different experiments, so we’ll rename it to that.</p></li>
<li><p>The airway dataset is from cell lines, and the parathyroid dataset is from people. We will introduce a “sample_type” covariate encoded with “cell_line” and “primary” values.</p></li>
<li><p>The rest of the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData()</a></code> columns will be dropped from both experiments.</p></li>
</ol>
<p>With these things in mind, let’s create new SummarizedExperiment objects, with harmonized <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData()</a></code>.</p>
<p><strong>airway</strong></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se.airway</span> <span class="op">&lt;-</span> <span class="va">airway</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">se.airway</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/eval.html">local</a></span><span class="op">(</span><span class="op">{</span>
  <span class="va">cd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">transmute</a></span><span class="op">(</span>
      sample_type <span class="op">=</span> <span class="st">"cell_line"</span>,
      cell_line <span class="op">=</span> <span class="va">cell</span>,
      treatment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">dex</span> <span class="op">==</span> <span class="st">"untrt"</span>, <span class="st">"control"</span>, <span class="st">"dex"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">DataFrame</span><span class="op">(</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cd</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">se.airway</span><span class="op">)</span>
  <span class="va">cd</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<p><strong>parathyroid</strong></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se.parathyroid</span> <span class="op">&lt;-</span> <span class="va">parathyroidGenesSE</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">se.parathyroid</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/eval.html">local</a></span><span class="op">(</span><span class="op">{</span>
  <span class="va">cd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">parathyroidGenesSE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">transmute</a></span><span class="op">(</span>
      sample_type <span class="op">=</span> <span class="st">"primary"</span>,
      subject_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"patient_"</span>, <span class="va">patient</span><span class="op">)</span>,
      treatment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html">tolower</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">treatment</span><span class="op">)</span><span class="op">)</span>,
      time <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"hrs"</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"h$"</span>, <span class="st">""</span>, <span class="va">time</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">DataFrame</span><span class="op">(</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cd</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">se.parathyroid</span><span class="op">)</span>
  <span class="va">cd</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="assay-data-assembly" class="section level3">
<h3 class="hasAnchor">
<a href="#assay-data-assembly" class="anchor"></a>Assay Data Assembly</h3>
<p>When you are assembling assay data together across different experiments, you need to decide which assays are the same, and which are different. For instance, if you have a series of experiments with paired-end RNA-seq data, and another set of experiments with single-end RNA-seq data: these aren’t the same assays, are they? … or?</p>
<p>You might decide to assemble these data together using two different assay types, one named <code>"gene_count_PE"</code> and the other <code>"gene_count_SE"</code>. There will be a disjoint set of samples that have assay data from one or the other. You then might decide to make a third assay type, <code>"gene_count_norm"</code> which somehow normalizes these data together and make that available for exploratory analysis.</p>
<p>In this example, for the sake of simplicity we will assume that the <code>airway</code> and <code>parathyroid</code> experiments were run on the same assay, and load them up together. This also means that they must share the same <em>feature space</em> for the assay, and we need to ensure that the <code><a href="https://rdrr.io/r/base/colnames.html">rownames()</a></code> of the two objects are equal.</p>
<p>Neither dataset provides gene information for its rows aside from the Ensembl identifiers. Because these were both quantitated against v75 (ish) of Ensembl’s gene annotations, we’ll iuse the <a href="http://bioconductor.org/packages/biomaRt">biomaRt</a> package to retrieve the information from the assembly close to that (<code><a href="https://rdrr.io/pkg/biomaRt/man/listEnsemblArchives.html">biomaRt::listEnsemblArchives()</a></code> points us to the URL we should query).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">loadNamespace</a></span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span>
<span class="va">mart</span> <span class="op">&lt;-</span> <span class="va">bm</span><span class="op">$</span><span class="fu">useMart</span><span class="op">(</span>
  host <span class="op">=</span> <span class="st">"feb2014.archive.ensembl.org"</span>,
  biomart <span class="op">=</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>,
  dataset <span class="op">=</span> <span class="st">"hsapiens_gene_ensembl"</span><span class="op">)</span>
<span class="va">mart.info</span> <span class="op">&lt;-</span> <span class="va">bm</span><span class="op">$</span><span class="fu">getBM</span><span class="op">(</span>
  attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ensembl_gene_id"</span>, <span class="st">"hgnc_symbol"</span>, <span class="st">"gene_biotype"</span><span class="op">)</span>,
  filters <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>,
  values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>,
  mart <span class="op">=</span> <span class="va">mart</span><span class="op">)</span></code></pre></div>
<p>The query above retrieves the data we need, but it can take a long time to run. To save time while compiling this vignette, we have serialized the result of the query, and will use that here.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mart.info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/eval.html">local</a></span><span class="op">(</span><span class="op">{</span>
  <span class="va">fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ensembl-v75-gene-info.csv.gz"</span>,
                    package <span class="op">=</span> <span class="st">"FacileData"</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html">gzfile</a></span><span class="op">(</span><span class="va">fn</span>, <span class="st">"rt"</span><span class="op">)</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<p>The column names of the <code>gene.info</code> table need to be tweaked, and we will only keep the rows (genes) that are shared between our two datasets.</p>
<div class="note">
<p>This mismatched gene matching wouldn’t be an issue if these data were from the same assay and processed in the same way. When you are assembling assay data into single assay of a FacileDataSet, the implicit assumption is that all of the data were processed in the same way.</p>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">shared.ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">se.airway</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">se.parathyroid</span><span class="op">)</span><span class="op">)</span>
<span class="va">gene.info</span> <span class="op">&lt;-</span> <span class="va">mart.info</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">transmute</a></span><span class="op">(</span>feature_id <span class="op">=</span> <span class="va">ensembl_gene_id</span>,
            feature_type <span class="op">=</span> <span class="st">"ensgid"</span>,
            name <span class="op">=</span> <span class="va">hgnc_symbol</span>, 
            meta <span class="op">=</span> <span class="va">gene_biotype</span>,
            source <span class="op">=</span> <span class="st">"Ensembl_v75"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">feature_id</span> <span class="op">%in%</span> <span class="va">shared.ids</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span><span class="op">(</span><span class="va">feature_id</span>, .keep_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">DataFrame</span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene.info</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gene.info</span><span class="op">[[</span><span class="st">"feature_id"</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p>Now we can cleanup the rowdata of our SummarizedExperiment and make them row-wise concordant.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se.airway</span> <span class="op">&lt;-</span> <span class="va">se.airway</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene.info</span><span class="op">)</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">se.airway</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gene.info</span>

<span class="va">se.parathyroid</span> <span class="op">&lt;-</span> <span class="va">se.parathyroid</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene.info</span><span class="op">)</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">se.parathyroid</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gene.info</span></code></pre></div>
</div>
</div>
<div id="faciledataset-creation" class="section level2">
<h2 class="hasAnchor">
<a href="#faciledataset-creation" class="anchor"></a>FacileDataSet Creation</h2>
<p>Like we’ve been saying all along … <em>all</em> you need is a named list of well-curated bioconductor experiments, and here we are!</p>
<p>We initialize a new FacileDataSet with our list of SummarizedExperiments (you can even just have one SummarizedExperiment), and specify which assay you are dropping in first.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se.all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>airway <span class="op">=</span> <span class="va">se.airway</span>, parathyroid <span class="op">=</span> <span class="va">se.parathyroid</span><span class="op">)</span>
<span class="va">xfds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.FacileDataSet.html">as.FacileDataSet</a></span><span class="op">(</span>
  <span class="va">se.all</span>,
  path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span><span class="op">)</span>,
  dataset_name <span class="op">=</span> <span class="st">"ExampleRnaFacileDataSet"</span>,
  assay_name <span class="op">=</span> <span class="st">"gene_counts"</span>,
  assay_description <span class="op">=</span> <span class="st">"Gene counts from their respective Bioconductor packages"</span>,
  assay_type <span class="op">=</span> <span class="st">"rnaseq"</span>,
  organism <span class="op">=</span> <span class="st">"Homo sapiens"</span><span class="op">)</span></code></pre></div>
</div>
<div id="tidy-like-data-retrieval" class="section level2">
<h2 class="hasAnchor">
<a href="#tidy-like-data-retrieval" class="anchor"></a>Tidy-like data retrieval</h2>
<p>Let’s look at how the correlation of TLN1 and ITGB1 change across datasets</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>

<span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">samples</span><span class="op">(</span><span class="va">xfds</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">with_assay_data</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ENSG00000137076"</span>, <span class="st">"ENSG00000150093"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/fetch_sample_covariates.html">with_sample_covariates</a></span><span class="op">(</span><span class="st">"treatment"</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">dat</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">ITGB1</span>, <span class="va">TLN1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">treatment</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dataset</span><span class="op">)</span></code></pre></div>
<p><img src="FacileDataSet-assembly_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bdat</span> <span class="op">&lt;-</span> <span class="fu">samples</span><span class="op">(</span><span class="va">xfds</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/fetch_assay_data.html">fetch_assay_data</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ENSG00000137076"</span>, <span class="st">"ENSG00000150093"</span><span class="op">)</span>,
                   normalized <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/fetch_sample_covariates.html">with_sample_covariates</a></span><span class="op">(</span><span class="st">"treatment"</span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">bdat</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">feature_name</span>, y <span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span>outlier.shape <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">treatment</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dataset</span><span class="op">)</span></code></pre></div>
<p><img src="FacileDataSet-assembly_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Steve Lianoglou, Vincent Rouilly, Peter Haverty, Denali Therapeutics, Genentech.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
